### PR TITLE
CPM-634: Add Get Product endpoint

### PR DESCRIPTION
**Pull request by pierallard**
Add a new API endpoint to access products through their UUID.

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Bundle/Controller/ExternalApi/GetProductByUuidController.php [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        } catch (ObjectNotFoundException $e) {
+            throw new NotFoundHttpException(sprintf('Product "%s" does not exist or you do not have permission to access it.', $uuid));
+        }
+
+        $normalizedProduct = $this->connectorProductWithUuidNormalizer->normalizeConnectorProduct($product);
+
+        return new JsonResponse($normalizedProduct);
+    }
+
+    private function temporaryGetUuidFromIdentifier(UuidInterface $uuid)
```
Comment (pierallard) [inline src/Akeneo/Pim/Enrichment/Bundle/Controller/ExternalApi/GetProductByUuidController.php:83]: This has to be removed once https://github.com/akeneo/pim-community-dev/pull/17367 is merged

File: src/Akeneo/Pim/Enrichment/Bundle/Controller/ExternalApi/GetProductByUuidController.php [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+            $this->getConnectorProductsWithOptions :
+            $this->getConnectorProducts;
+
+        try {
+            $user = $this->tokenStorage->getToken()->getUser();
+            Assert::isInstanceOf($user, UserInterface::class);
+
+            $identifier = $this->temporaryGetUuidFromIdentifier(Uuid::fromString($uuid));
+
+            $product = $connectorProductsQuery->fromProductIdentifier($identifier, $user->getId());
```
Comment (pierallard) [inline src/Akeneo/Pim/Enrichment/Bundle/Controller/ExternalApi/GetProductByUuidController.php:63]: This has to be "fromProductUuid" directly

File: src/Akeneo/Pim/Enrichment/Bundle/Controller/ExternalApi/GetProductByUuidController.php [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        $connectorProductsQuery = 'true' === $request->query->get('with_attribute_options', 'false') ?
+            $this->getConnectorProductsWithOptions :
+            $this->getConnectorProducts;
+
+        try {
+            $user = $this->tokenStorage->getToken()->getUser();
+            Assert::isInstanceOf($user, UserInterface::class);
+
+            $uuid = Uuid::fromString($uuid);
+            if ($uuid->getVersion() !== 4) {
```
Comment (pierallard) [inline src/Akeneo/Pim/Enrichment/Bundle/Controller/ExternalApi/GetProductByUuidController.php:59]: remove this

File: src/Akeneo/Pim/Enrichment/Bundle/Controller/ExternalApi/GetProductByUuidController.php [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+            $product = $connectorProductsQuery->fromProductUuid($uuid, $user->getId());
+            $this->eventDispatcher->dispatch(new ReadProductsEvent(1));
+
+            if ($request->query->getAlpha('with_quality_scores', 'false') === 'true') {
+                $product = $this->getProductsWithQualityScores->fromConnectorProduct($product);
+            }
+            if ($request->query->getAlpha('with_completenesses', 'false') === 'true') {
+                $product = $this->getProductsWithCompletenesses->fromConnectorProduct($product);
+            }
+        } catch (InvalidUuidStringException) {
```
Comment (pierallard) [inline src/Akeneo/Pim/Enrichment/Bundle/Controller/ExternalApi/GetProductByUuidController.php:72]: remove this and in the Delete too

File: tests/back/Pim/Enrichment/EndToEnd/Product/Product/ExternalApi/GetProductWithUuidEndToEnd.php [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+        $content = json_decode($response->getContent(), true);
+        $this->assertCount(2, $content, 'response contains 2 items');
+        $this->assertSame(Response::HTTP_NOT_FOUND, $content['code']);
+        $this->assertSame(
+            sprintf('Product "%s" does not exist or you do not have permission to access it.', $uuid),
+            $content['message']
+        );
+    }
+
+    public function test_it_throws_a_400_with_a_uuid1()
```
Comment (pierallard) [inline tests/back/Pim/Enrichment/EndToEnd/Product/Product/ExternalApi/GetProductWithUuidEndToEnd.php:226]: to remove

**Review by pierallard — COMMENTED** [review]
mobe revioued

**Review by Balamung — APPROVED** [review]

**Review by cemorin — APPROVED** [review]

**Pull request merged**